src/register/reg_crdt.rs

```diff
@@ -0,0 +1,139 @@
+#[derive(Clone, Serialize, Deserialize, PartialEq, Eq, Hash, PartialOrd)]
+pub struct RegisterCrdt {
+    /// Address on the network of this piece of data
+    address: Address,
+    /// CRDT to store the actual data, i.e. the items of the Register.
+    data: MerkleReg<Entry>,
+}
+
+impl Display for RegisterCrdt {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "(")?;
+        for (i, entry) in self.data.read().values().enumerate() {
+            if i > 0 {
+                write!(f, ", ")?;
+            }
+            write!(f, "<{}>", String::from_utf8_lossy(&entry),)?;
+        }
+        write!(f, ")")
+    }
+}
+
+impl RegisterCrdt {
+    /// Constructs a new 'RegisterCrdt'.
+    pub fn new(address: Address) -> Self {
+        Self {
+            address,
+            data: MerkleReg::new(),
+        }
+    }
+
+    /// Returns the address.
+    pub fn address(&self) -> &Address {
+        &self.address
+    }
+
+    /// Returns total number of items in the register.
+    pub fn size(&self) -> u64 {
+        (self.data.num_nodes() + self.data.num_orphans()) as u64
+    }
+
+    /// Write a new entry to the RegisterCrdt, returning the hash
+    /// of the entry and the CRDT operation without a signature
+    pub fn write(
+        &mut self,
+        entry: Entry,
+        parents: BTreeSet<EntryHash>,
+        source: PublicKey,
+    ) -> Result<(EntryHash, CrdtOperation<Entry>)> {
+        let address = *self.address();
+
+        let crdt_op = self.data.write(entry, parents);
+        self.data.apply(crdt_op.clone());
+        let hash = crdt_op.hash();
+
+        // We return the operation as it may need to be broadcasted to other replicas
+        let op = CrdtOperation {
+            address,
+            crdt_op,
+            source,
+            signature: None,
+        };
+
+        Ok((hash, op))
+    }
+
+    /// Apply a remote data CRDT operation to this replica of the RegisterCrdt.
+    pub fn apply_op(&mut self, op: CrdtOperation<Entry>) -> Result<()> {
+        // Let's first check the op is validly signed.
+        // Note: Perms for the op are checked at the upper Register layer.
+        let sig = op.signature.ok_or(Error::CrdtMissingOpSignature)?;
+        let bytes_to_verify = utils::serialise(&op.crdt_op).map_err(|err| {
+            Error::Serialisation(format!(
+                "Could not serialise CRDT operation to verify signature: {}",
+                err
+            ))
+        })?;
+        op.source.verify(&sig, &bytes_to_verify)?;
+
+        // Apply the CRDT operation to the Register
+        self.data.apply(op.crdt_op);
+
+        Ok(())
+    }
+
+    /// Gets the entry corresponding to the provided `hash` if it exists.
+    pub fn get(&self, hash: EntryHash) -> Option<&Entry> {
+        self.data.node(hash).map(|node| &node.value)
+    }
+
+    /// Gets the last entry, or entries if there are branches.
+    pub fn last(&self) -> BTreeSet<(EntryHash, Entry)> {
```

Comment:
The Register CRDT family has a few common interfaces, we should follow one of them instead of using our own concepts here (i.e. `last`):

The one rust-crdt uses:
```
Register:
  write(values, ctx)
  read()
```

Other common ones:
```
Register:
  update(val, ctx)
  query()
```
```
Register:
  assign(val, ctx)
  value()
```

I'm not religious about which one to pick, but I'd prefer we follow one of the common conventions.